akka-actor/src/main/java/akka/dispatch/AbstractNodeQueue.java

```diff
@@ -85,18 +144,23 @@ public final T poll() {
         }
     }
     
-    /*
+    /**
+     * Pull one item from the queue, returning it within a queue node.
+     * 
      * Use this method only from the consumer thread!
+     * 
+     * @return queue node with element inside if there was one, or null if there was none
      */
     @SuppressWarnings("unchecked")
     public final Node<T> pollNode() {
-      Node<T> tail;
-      Node<T> next;
-      for(;;) {
-        tail = ((Node<T>)Unsafe.instance.getObjectVolatile(this, tailOffset));
-        next = tail.next();
-        if (next != null || get() == tail)
-          break;
+      final Node<T> tail = (Node<T>) Unsafe.instance.getObjectVolatile(this, tailOffset);
+      Node<T> next = tail.next();
+      if (next == null && get() != tail) {
+          // if tail != head this is not going to change until producer makes progress
+          // we can avoid reading the head and just spin on next until it shows up
+          do {
+              next = tail.next();
```

Comment:
Perhaps add a TODO to use http://mail.openjdk.java.net/pipermail/hotspot-compiler-dev/2016-January/020987.html here?